File: tools/ghidra-scripts/ExportAnalysis.java
// Ghidra GhidraScript: exports per-binary analysis artifacts.
// Outputs to notes/binary-analysis/<prog-name>/
//   ghidra-functions.txt    — all functions with addresses + sizes
//   ghidra-decompiled.c     — decompiled C for security-relevant functions
//
// Run via analyzeHeadless ... -process <binary> -postScript ExportAnalysis.java

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.util.task.ConsoleTaskMonitor;

import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;

public class ExportAnalysis extends GhidraScript {

    private static final String[] SECURITY_KEYWORDS = {
        "root", "auth", "token", "authoriz", "permission", "privilege",
        "reboot", "update", "ota", "upgrade", "password", "passwd",
        "cmd_", "unlock", "enable", "disable", "verify",
        "sign", "crypt", "key", "cert", "tls", "ssl",
    };

    @Override
    public void run() throws Exception {
        String base = "/home/anthony/Documents/Hacking/Creality-K1c-Printer";
        String progName = currentProgram.getName();
        String outDir = base + "/notes/binary-analysis/" + progName;

        File dir = new File(outDir);
        if (!dir.exists()) dir.mkdirs();

        println("[ExportAnalysis] Program: " + progName);
        println("[ExportAnalysis] Output:  " + outDir);

        // Export all functions
        File fnFile = new File(outDir + "/ghidra-functions.txt");
        int fnCount = 0;
        try (PrintWriter pw = new PrintWriter(new FileWriter(fnFile))) {
            FunctionIterator fns = currentProgram.getFunctionManager()
                                                 .getFunctions(true);
            while (fns.hasNext()) {
                Function fn = fns.next();
                pw.printf("%s\t%d\t%s%n",
                    fn.getEntryPoint(),
                    fn.getBody().getNumAddresses(),
                    fn.getName());
                fnCount++;
            }
        }
        println("[ExportAnalysis] Exported " + fnCount + " functions");

        // Decompile security-relevant functions
        DecompInterface decomp = new DecompInterface();
        decomp.openProgram(currentProgram);
        ConsoleTaskMonitor monitor = new ConsoleTaskMonitor();

        File dcFile = new File(outDir + "/ghidra-decompiled.c");
        int dcCount = 0;
        try (PrintWriter pw = new PrintWriter(new FileWriter(dcFile))) {
            FunctionIterator fns = currentProgram.getFunctionManager()
                                                 .getFunctions(true);
            while (fns.hasNext()) {
                Function fn = fns.next();
                String lower = fn.getName().toLowerCase();
                boolean match = false;
                for (String kw : SECURITY_KEYWORDS) {
                    if (lower.contains(kw)) { match = true; break; }
                }
                if (!match) continue;

                DecompileResults res = decomp.decompileFunction(fn, 90, monitor);
                if (res != null && res.decompileCompleted()) {
                    ghidra.app.decompiler.DecompiledFunction dc =
                        res.getDecompiledFunction();
                    if (dc != null) {
                        pw.println("\n// === " + fn.getName()
                                 + " @ " + fn.getEntryPoint() + " ===");
                        pw.println(dc.getC());
                        dcCount++;
                    }
                }
            }
        }
        println("[ExportAnalysis] Decompiled " + dcCount
              + " security-relevant functions → " + dcFile.getPath());
        println("[ExportAnalysis] Done.");
    }
}

File: tools/ghidra-scripts/ExportAnalysis.py
# Ghidra headless post-analysis script (Jython/Python 2.7 compatible).
# Exports per-binary analysis artifacts to notes/binary-analysis/<name>/:
#   ghidra-functions.txt   — all functions with addresses
#   ghidra-decompiled.c    — decompiled C for security-relevant functions
#
# Run via:
#   analyzeHeadless <project> K1C-2025 -process <binary> \
#     -postScript ExportAnalysis.py -scriptPath tools/ghidra-scripts/

import os
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

BASE = "/home/anthony/Documents/Hacking/Creality-K1c-Printer"
OUT_ROOT = os.path.join(BASE, "notes", "binary-analysis")

SECURITY_KEYWORDS = [
    "root", "auth", "token", "authoriz", "permission", "privilege",
    "reboot", "update", "ota", "upgrade", "password", "passwd",
    "CMD_", "cmd_", "unlock", "enable", "disable", "verify",
    "sign", "crypt", "key", "cert", "tls", "ssl",
]

prog_name = currentProgram.getName()  # noqa: F821 (Ghidra API global)
out_dir = os.path.join(OUT_ROOT, prog_name)
if not os.path.exists(out_dir):
    os.makedirs(out_dir)

print("[ExportAnalysis] Program: {}".format(prog_name))
print("[ExportAnalysis] Output:  {}".format(out_dir))

# --- Export all function addresses + names ---
fn_manager = currentProgram.getFunctionManager()  # noqa: F821
fn_list_path = os.path.join(out_dir, "ghidra-functions.txt")
count = 0
with open(fn_list_path, "w") as f:
    for fn in fn_manager.getFunctions(True):
        f.write("{}\t{}\n".format(fn.getEntryPoint(), fn.getName()))
        count += 1
print("[ExportAnalysis] Exported {} functions to {}".format(count, fn_list_path))

# --- Decompile security-relevant functions ---
decomp = DecompInterface()
decomp.openProgram(currentProgram)  # noqa: F821
monitor = ConsoleTaskMonitor()

decompiled_path = os.path.join(out_dir, "ghidra-decompiled.c")
decompiled_count = 0
with open(decompiled_path, "w") as f:
    for fn in fn_manager.getFunctions(True):
        fn_name_lower = fn.getName().lower()
        if any(kw.lower() in fn_name_lower for kw in SECURITY_KEYWORDS):
            result = decomp.decompileFunction(fn, 90, monitor)
            if result.decompileCompleted():
                dc = result.getDecompiledFunction()
                if dc is not None:
                    f.write("\n// === {} @ {} ===\n".format(
                        fn.getName(), fn.getEntryPoint()))
                    f.write(dc.getC())
                    decompiled_count += 1

print("[ExportAnalysis] Decompiled {} security-relevant functions".format(decompiled_count))
print("[ExportAnalysis] Output: {}".format(decompiled_path))
print("[ExportAnalysis] Done.")
